File: src/c/magma/FirstLocator.c
import magma.option.None;
import magma.option.Option;
import magma.option.Some;
struct FirstLocator(String infix) implements Locator {@Override
    public Option<Integer> locate(String input) {
        final var index = input.indexOf(infix());
        if (index == -1) return new None<>();
        return new Some<>(index);
    }@Override
    public int sliceLength() {
        return this.infix.length();
    }
}

File: src/java/magma/FirstLocator.java
package magma;

import java.util.Optional;

public record FirstLocator(String slice) implements Locator {
    @Override
    public Optional<Integer> locate(String input) {
        final var index = input.indexOf(this.slice);
        if (index == -1) return Optional.empty();
        return Optional.of(index);
    }

    @Override
    public int length() {
        return this.slice.length();
    }
}